testpilot-metrics.js

```diff
@@ -0,0 +1,274 @@
+          object: this.id
+        }
+      };
+
+      try {
+        Services.obs.notifyObservers(subject, 'testpilot::send-metric', stringified);
+        this._log(`Sent client message via nsIObserverService: ${stringified}`);
+      } catch (ex) {
+        this._error(`Failed to send nsIObserver client ping: ${ex}`);
+        return;
+      }
+    }
+  },
+
+  /**
+   * Sends an event to Google Analytics, using sendBeacon and the Measurement
+   * Protocol API.
+   * @private
+   * @param {string} event - see `sendEvent` docs
+   * @param {string} object - see `sendEvent` docs
+   * @param {string} category - see `sendEvent` docs. Note that `category` is
+   * required here, assuming the default value was filled in by `sendEvent`.
+   * @param {string} [study] - see `sendEvent` docs
+   * @param {string} [variant] - see `sendEvent` docs
+   */
+  _sendToGA: function(event, object, category, study, variant) {
+    const data = {
+      v: 1,
+      tid: this.tid,
+      cid: this.cid,
+      t: 'event',
+      ec: category,
+      ea: object,
+      el: event
+    };
+    if (study && variant) {
+      data.xid = study;
+      data.xvar = variant;
+    }
+    const encoded = this._formEncode(data);
+    const GA_URL = 'https://ssl.google-analytics.com/collect';
+    if (this.type === 'webextension') {
+      navigator.sendBeacon(GA_URL, encoded);
+    } else {
+      // SDK and bootstrapped types might not have a window reference, so get
+      // the sendBeacon DOM API from the hidden window.
+      Services.appShell.hiddenDOMWindow.navigator.sendBeacon(GA_URL, encoded);
+    }
+    this._log(`Sent GA message: ${encoded}`);
+  },
+
+  /**
+   * Serializes an object into x-www-form-urlencoded format.
+   * @example
+   * // returns 'a=b&foo=b+ar'
+   * metrics._formEncode({a: 'b', foo: 'b ar'});
+   * @private
+   * @param {Object} obj - Any JS object
+   * @returns {string}
+   */
+  _formEncode: function(obj) {
+    const params = [];
+    if (!obj) { return ''; }
+    Object.keys(obj).forEach(item => {
+      let encoded = encodeURIComponent(item) + '=' + encodeURIComponent(obj[item]);
+      encoded = encoded.replace(/%20/g, '+');
+      params.push(encoded);
+    });
+    return params.join('&');
+  },
+
+  /**
+   * Initializes transports used for sending messages. For WebExtensions,
+   * creates a `BroadcastChannel` (transport for client pings). WebExtensions
+   * use navigator.sendBeacon for GA transport, and they always have access
+   * to DOM APIs, so there's no setup work required. For other types, loads
+   * `Services.jsm`, which exposes the nsIObserverService (transport for client
+   * pings), and exposes the navigator.sendBeacon API (GA transport) via the
+   * appShell service's hidden window.
+   * @private
+   * @throws {Error} if transport setup unexpectedly fails
+   */
+  _initTransports: function() {
+    if (this.type === 'webextension') {
+      try {
+        this._channel = new BroadcastChannel(this.topic);
```

Comment:
Potential problem in WebExtension land here: [The BroadcastChannel has to be named `testpilot-telemetry`](https://github.com/mozilla/testpilot/blob/master/addon/lib/webextension-channels.js#L17), but it looks like it will be named just `testpilot` because [`this.topic` defaults to that](https://github.com/mozilla/testpilot-metrics/pull/4/files#diff-d2211d7ba9ac51d585a2fb0a21aaa056R40).